lib/Corres_Method.thy

```diff
@@ -0,0 +1,212 @@
+    corres_pre0
+    (* fix up schematic guards if they contain constructor parameters *)
+    | wpfix
+    (* apply a single corres rule if possible *)
+    | corres_rule m simp: term_simp simp simp_del: simp_del split_del: split_del split: split
+                    cong: cong corres: corres corres_term: corres_term
+    (* only split if we can apply a single corres rule afterwards *)
+    | corres_split corres_splits: corres_splits,
+      corres_rule m simp: simp term_simp simp_del: simp_del split_del: split_del split: split
+                    cong: cong corres: corres corres_term: corres_term
+    (* apply potentially unsafe fallback rules if any are provided *)
+    | corres_rule m simp: simp term_simp simp_del: simp_del split_del: split_del split: split
+                    cong: cong corres: fallback
+    (* simplify head corres goal, e.g. for things like liftM unfolding if the user provides such
+       a rule as "simp". Not clarsimp, because clarsimp will still perform hypsubst on assumptions
+       and might through that rewrite guards *)
+    | succeeds \<open>is_corres\<close>,
+      simp cong: corres_weaker_cong cong split: split split del: if_split split_del
+           add: simp corres_simp del: corres_no_simp simp_del
+    (* simplify any remaining side condition head goal (non-corres, non-wp). This is either a side
+       condition that was not solved by corres_cleanup, or it is one of the two terminal implication
+       goals. It is very likely that the method will stop after this and not have solved the goal,
+       but it also very likely that the first thing we want to do for such a goal is clarsimp. That
+       means, overall we might solve a few more goals, and not be detrimental to interactive proof
+       either. *)
+    | fails \<open>is_corres\<close>, fails \<open>is_wp\<close>,
+      clarsimp cong: cong simp del: simp_del simp: simp split del: if_split split_del split: split
+               intro!: intro
+    (* if (and only if) we get to the state where all corres goals and side conditions are solved,
+       attempt to solve all wp goals that were generated in order. We are not using then_all_new_fwd
+       here, because we should only start solving wp goals once *all* corres goals are solved,
+       because otherwise the goal will still have schematic post conditions. Fail if there is a
+       free schematic postcondition despite all these measures.
+       *)
+    | succeeds \<open>is_wp\<close>, fails \<open>is_hoare_schematic_post\<close>,
+      (wpsimp wp: wp wp_del: wp_del simp: simp simp_del: simp_del split: split split_del: split_del
+              cong: cong)+
+   )+)[1]
+
+(* Instance of the corres' method with default cleanup tactic *)
+method corres
+  uses simp term_simp simp_del split split_del cong intro wp wp_del fallback
+  declares corres corres_term corres_splits =
+  corres' \<open>fail\<close> simp: simp term_simp: term_simp simp_del: simp_del split: split
```

Comment:
does this mean the default cleanup tactic is `fail`?

Reply:
Ah wait, I think I get it, that's passed all the way down ... yeah, would still be useful to hint at why you want `fail` here in a/the comment.

Reply:
good point. `succeed` would always stop the cleanup and do nothing, `fail` lets all other tactics work.